### ISSUE TITLE
Dataclass not recognized if created by custom decorator

### ISSUE THREAD

**Issue by huwcbjones** [PyCQA/pylint#7226]
### Bug description

When decorating a class with a decorator that returns the class wrapped in a dataclass (see MWE below), pylint fails to realise this new class is a dataclass.

The example below causes pylint to think that `HelloMsg` is not a dataclass, therefore it wrongly thinks that the `msg` kwarg is invalid.

TL;DR: dynamic dispatch and laziness to keep adding the dataclass decorator
```py
# pylint: disable=missing-docstring
# pylint: disable=too-few-public-methods
from dataclasses import dataclass


def dec(tag: str):
    def _wrapper(cls):
        setattr(cls, "__tag__", tag)
        return dataclass(cls)
    return _wrapper


@dataclass
class Message:
    id: int


@dec("hello")
class HelloMsg(Message):
    msg: str


msg = HelloMsg(id=1, msg="hello")
```

This example works as expected
```py
# pylint: disable=missing-docstring
# pylint: disable=too-few-public-methods
from dataclasses import dataclass


def dec(tag: str):
    def _wrapper(cls):
        setattr(cls, "__tag__", tag)
        return cls
    return _wrapper


@dataclass
class Message:
    id: int


@dec("hello")
@dataclass
class HelloMsg(Message):
    msg: str


msg = HelloMsg(id=1, msg="hello")
```

### Command used

```shell
pylint mwe.py
```

### Pylint output

```shell
************* Module mwe
mwe.py:23:6: E1123: Unexpected keyword argument 'msg' in constructor call (unexpected-keyword-arg)
```

### Expected behavior

No message to be emitted

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.9.1 (default, Mar 15 2021, 18:22:18)
[Clang 12.0.0 (clang-1200.0.32.27)]
```

### OS / Environment

macOS 11.6.1

**Comment by jacobtylerwalls** [comment]
The check for dataclasses is profoundly dumb (and I mean that neutrally and literally :D): it checks for the literal name of the decorator, see:
https://github.com/PyCQA/astroid/blob/b5812cc97bc7d63b72e2a99b7060a253d5ca3ac7/astroid/brain/brain_dataclasses.py#L42-L46
